### DISCUSSION TITLE
Is there some way to reference an externally built library project?

### DISCUSSION THREAD

**Discussion by pmunts** [alire-project/alire#1306] [Q&A]
I would like to somehow use the prebuilt Debian 11 AWS library (only in conjunction with the Debian 11 compiler, of course).

I have confirmed that copying `/usr/share/gpr/aws.gpr` to `config/` and adding `with "aws.gpr";` to `config/xxxx_config.gpr` works perfectly.

Is it possible to define a crate that does not contain any source code, just a `.gpr` file for an externally built library project?

**Comment by pmunts** [comment]
To answer my own question, yes, you can create a crate containing nothing but aws.gpr.  It works perfectly for using the AWS library in the Debian distribution.

**Comment by mosteo** [comment]
I'm turning this into a discussion.

Isn't what you ask essentially what an external crate definition will get you?

Also note that Ada packages in Debian are expected to be used with the system compiler; otherwise .ali files won't match and try to cause recompilations for sources that are unavailable (IIRC from experiences long ago).

**Reply by pmunts** [reply]
Yes, that is exactly what I want:  to use the Debian 11 system AWS library with Debian 11 system compiler, in an Alire project, just as I already do outside Alire.  AWS is a monster to build (as the arcane procedure in the aws crate illustrates) and practically impossible to cross-compile.  Under the _best_ of circumstances it takes a long time to build.

So I submitted a pull request for 6 new aws_* crates, one for Debian 11 and 5 for the compilers in the MuntsOS Embedded Linux cross-toolchains.  Since Alire still has no Linux target cross-compilers, I predict it will not in any useful time frame provide a solution for AWS for MuntsOS, except the crates I submitted.  Although I was able to build and test them locally, they failed the Alire build checks because I am (as I well know) torturing Alire in ways you guys never imagined.

Nonetheless, those 6 new aws_* crates do work, and are useful, and I would appreciate it if you would merge them.  They are all named such that it is highly unlikely anyone will try to use them unless they know what they are doing (probability of harm is very low).

**Reply by Fabien-Chouteau** [reply]
It's the `aws.gpr` files provided by your distribution that you should `with` in from your main GPR file. There is no need for those copies in crates.

I agree that in an ideal world the `aws` Alire crate should be able to cross compile, but it's not possible right now. Note that this is because of the way AWS is built (using configure and Makefile).
